### DISCUSSION TITLE
Adding a Validation utility method to help validate the MachineConfig

### DISCUSSION THREAD

**Discussion by computnik** [statelyai/xstate#1537] [Ideas]
There might be use-cases where the xstate library is served behind an API/Service. In such cases, where new statecharts can be authored on the fly, it might be really helpful to have a `validateMachineConfig` util method.

### Possible implementation
This can be potentially done using the existing [machine.schema.json](https://github.com/davidkpiano/xstate/blob/master/packages/core/src/machine.schema.json), like shown in [json.test.ts](https://github.com/davidkpiano/xstate/blob/master/packages/core/test/json.test.ts)

Potentially this can be achieved by adding a file like `core/src/validateMachineConfig.ts`, probably with implementation, on the lines of the following snippet

```typescript
import Ajv from 'ajv';
import { createMachine } from './index';
import * as machineSchema from './machine.schema.json';

const ajv = new Ajv();
const validate = ajv.compile(machineSchema);

export const validateMachineConfig = (config: MachineConfig) => {
  const serializedDefinition = JSON.parse(JSON.stringify(machine.definition));
  const ajv = new Ajv();
  const valid = ajv.validate(machineSchema, config);
  if (!valid) {
    throw new Error(ajv.errors.join(',')) // or can return false with errors
  }
  return true
}

```

Let me know if you think it might be of value. Also, would love to contribute towards this too!

**Comment by davidkpiano** [comment]
This would definitely be of value. My perspective on validation includes JSON Schema validation, but goes beyond that - we also want to check that there are no "orphaned" states with no transitions to it, that there are no "sink" states with no outgoing transitions, that parallel states have more than one "region" (child state), etc.

**Comment by computnik** [comment]
@davidkpiano that sounds like a great idea, and I do think having all of these validations would be amazing. 
Also, in that case, would it also make sense to also associate WARNINGS vs ERROR scenarios with these validations, like for `Orphaned` states, etc. where it might not break the system, but it "might" be a mistake.
